src/messenger/messenger.cpp

```diff
@@ -398,128 +404,225 @@ void Messenger::Terminate() {
-                            const std::string &message, CallbackFn callback, uint64_t remote_cb_id) {
-  uint64_t send_msg_id = GetNextSendMessageId();
+  // Build and queue the message to be sent.
+  common::ManagedPointer<zmq::socket_t> socket = common::ManagedPointer(connection->socket_);
+  {
+    std::unique_lock lock(pending_messages_mutex_);
+    pending_messages_.emplace(
+        msg_id,
+        PendingMessage{socket, connection->target_name_,
+                       ZmqMessage::Build(msg_id, sender_cb_id, remote_cb_id, connection->routing_id_, message), false});
+  }
+}
 
-  // Register the callback that will be invoked when a response to this message is received.
-  callbacks_mutex_.lock();
-  callbacks_[send_msg_id] = std::move(callback);
-  callbacks_mutex_.unlock();
+void Messenger::SendMessage(const router_id_t router_id, const std::string &recv_id, const std::string &message,
+                            CallbackFn callback, callback_id_t remote_cb_id) {
+  common::ManagedPointer<ConnectionRouter> router = common::ManagedPointer(routers_.at(router_id));
+  message_id_t msg_id = next_message_id_++;
+  callback_id_t send_cb_id = GetBuiltinCallback(BuiltinCallback::NOOP);
+  if (callback != nullptr) {
+    // Register the callback that will be invoked when a response to this message is received.
+    send_cb_id = GetNextSendCallbackId();
+    callbacks_mutex_.lock();
+    callbacks_[send_cb_id] = std::move(callback);
+    callbacks_mutex_.unlock();
+  }
 
   // Build and send the message. Note that ConnectionRouter is a ROUTER socket.
-  zmq::message_t router_data(recv_id.data(), recv_id.size());
-  if (router_id->socket_->send(router_data, zmq::send_flags::sndmore).has_value()) {
-    ZmqMessage reply = ZmqMessage::Build(send_msg_id, remote_cb_id, router_id->identity_, message);
-    ZmqUtil::SendMsgIdentity(common::ManagedPointer(router_id->socket_.get()), router_id->identity_);
-    ZmqUtil::SendMsgPayload(common::ManagedPointer(router_id->socket_.get()), reply);
-    MESSENGER_LOG_TRACE(fmt::format("[PID={}] Messenger ({}) SENT-TO {}: {} ", ::getpid(), router_id->identity_,
-                                    recv_id, reply.GetRawPayload()));
+  common::ManagedPointer<zmq::socket_t> socket = common::ManagedPointer(router->socket_);
+  {
+    std::unique_lock lock(pending_messages_mutex_);
+    pending_messages_.emplace(
+        msg_id, PendingMessage{socket, recv_id,
+                               ZmqMessage::Build(msg_id, send_cb_id, remote_cb_id, router->identity_, message), true});
   }
 }
 
-common::ManagedPointer<ConnectionRouter> Messenger::GetConnectionRouter(const std::string &router_id) {
-  return common::ManagedPointer(routers_.at(router_id).get());
+callback_id_t Messenger::GetNextSendCallbackId() {
+  callback_id_t send_cb_id = next_callback_id_++;
+  // Check for wraparound.
+  if (0 == next_callback_id_.load().UnderlyingValue()) {
```

Comment:
Does it matter that this `if` block is not necessarily atomic? When `next_callback_id_.store` is called the value isn't necessarily still 0. It would be nice if there was a test and set operation we could use, `compare-exchange` (https://en.cppreference.com/w/cpp/atomic/atomic/compare_exchange) seems like it might be the same thing?

Reply:
yep you're right!